pkg/controller/tenant/tenant_controller.go

```diff
@@ -0,0 +1,198 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/util/metrics"
+
+	"k8s.io/klog"
+)
+
+var tenant_defult_namespaces = [...]string{"default", "kube-system", "kube-public"}
+
+// tenantController is responsible for performing actions dependent upon a tenant phase
+type tenantController struct {
+	// lister that can list tenants from a shared cache
+	lister corelisters.TenantLister
+	// returns true when the tenant cache is ready
+	listerSynced cache.InformerSynced
+	// tenants that have been queued up for processing by workers
+	queue workqueue.RateLimitingInterface
+	// kubeclient for api calls
+	kubeClient clientset.Interface
+	// sync handler for injection
+	syncHandler func(key string) error
+}
+
+// NewTenantController creates a new iinstance of tenantcontroller
+func NewTenantController(
+	kubeClient clientset.Interface,
+	tenantInformer coreinformers.TenantInformer,
+	resyncPeriod time.Duration) *tenantController {
+
+	// create the controller so we can inject the enqueue function
+	tenantController := &tenantController{
+		kubeClient: kubeClient,
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "tenant"),
+	}
+
+	if kubeClient != nil && kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		metrics.RegisterMetricAndTrackRateLimiterUsage("tenant_controller", kubeClient.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	// configure the tenant informer event handlers
+	tenantInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				tenantController.enqueue(obj)
+			},
+		},
+		resyncPeriod,
+	)
+	tenantController.lister = tenantInformer.Lister()
+	tenantController.listerSynced = tenantInformer.Informer().HasSynced
+	tenantController.syncHandler = tenantController.syncTenant
+
+	return tenantController
+}
+
+// Run starts the controller with the specified number of workers.
+func (tc *tenantController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer tc.queue.ShutDown()
+
+	klog.Infof("Starting tenant controller.")
+	defer klog.Infof("Shutting down tenant controller")
+
+	if !controller.WaitForCacheSync("tenant", stopCh, tc.listerSynced) {
+		return
+	}
+
+	klog.V(5).Info("Starting workers of tenant controller")
+	for i := 0; i < workers; i++ {
+		go wait.Until(tc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// worker processes the queue of tenant objects.
+func (tc *tenantController) worker() {
+	workFunc := func() bool {
+		key, quit := tc.queue.Get()
+		if quit {
+			return true
+		}
+		defer tc.queue.Done(key)
+
+		err := tc.processQueue(key.(string))
+		if err == nil {
+			// no error, forget this entry and return
+			tc.queue.Forget(key)
+			return false
+		} else {
+			// rather than wait for a full resync, re-add the tenant to the queue to be processed
```

Comment:
Instead of just adding this key back to queue, I prefer a max-retries strategy.

Reply:
The current way seems a common pattern being followed by lots of controllers. I'd like to first finish the functionality following the current patterns, and then let's see how this controller and other controllers can be improved.  I created issue #131 to track this.